_posts/2020-12-30-fast-jdbc-revisited.md

```diff
@@ -0,0 +1,252 @@
+Additionally, we have added the `jpype.dbapi2` module as a third possibility to access databases with a JDBC driver from Python.
+
+To have a single JAR that we can use to start JVM as in the initial post, we update the `pom.xml` to the new version of the dependencies and build that JAR using `mvn assembly:single`.
+
+```xml
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>com.uwekorn</groupId>
+    <artifactId>drill-odbc</artifactId>
+    <version>0.2-SNAPSHOT</version>
+    <packaging>jar</packaging>
+
+    <name>drill-odbc</name>
+
+    <properties>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    </properties>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.arrow</groupId>
+            <artifactId>arrow-jdbc</artifactId>
+            <version>2.0.0</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.arrow</groupId>
+            <artifactId>arrow-memory</artifactId>
+            <version>2.0.0</version>
+            <type>pom</type>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.drill.exec</groupId>
+            <artifactId>drill-jdbc-all</artifactId>
+            <version>1.18.0</version>
+            <exclusions>
+              <exclusion>
+                <groupId>org.slf4j</groupId>
+                <artifactId>log4j-over-slf4j</artifactId>
+              </exclusion>
+        </exclusions>
+        </dependency>
+
+    </dependencies>
+    
+     <build>
+      <plugins>
+        <plugin>
+          <artifactId>maven-assembly-plugin</artifactId>
+          <configuration>
+            <archive>
+              <manifest>
+                <mainClass>com.uwekorn.Main</mainClass>
+              </manifest>
+            </archive>
+            <descriptorRefs>
+              <descriptorRef>jar-with-dependencies</descriptorRef>
+            </descriptorRefs>
+          </configuration>
+        </plugin>
+          <plugin>
+              <groupId>org.apache.maven.plugins</groupId>
+              <artifactId>maven-compiler-plugin</artifactId>
+              <version>3.8.0</version>
+              <configuration>
+                  <source>11</source>
+                  <target>11</target>
+              </configuration>
+          </plugin>
+      </plugins>
+    </build>
+</project>
+```
+
+We then start the JVM as before using the following code.
+
+```python
+import jpype
+import os
+
+classpath = os.path.join(os.getcwd(), "all-jar/target/drill-odbc-0.2-SNAPSHOT-jar-with-dependencies.jar")
+jpype.startJVM(jpype.getDefaultJVMPath(), f"-Djava.class.path={classpath}")
```

Comment:
This is still using a fairly old pattern.   Directly setting up the classpath is still acceptable, but has the risk that if you need more than once class element you need to use the system separator which is different between windows and linux.

I would recommend the the code be changed to 
```
classpath = os.path.join(os.getcwd(), "all-jar/target/drill-odbc-0.2-SNAPSHOT-jar-with-dependencies.jar")
jpype.startJVM(classpath=classpath)
```

The classpath argument can either be a string or a list of strings.  Wildcards are accepted.  This will automatically handle the separator if more than one is provided.  

The other option is to use addClassPath.  You can call addClassPath either before or after the startJVM statement which is great if you already have the JVM open and want to then add the library to it after.  There may still be some interactions on this as I haven't tested it extensively with jdbc.

Reply:
`addClassPath` didn't work after the JVM was started. Then the driver couldn't be found, I'll open an issue with a reproducible example.

Reply:
https://github.com/jpype-project/jpype/issues/914